### ISSUE TITLE
UMAP eating up memory

### ISSUE THREAD

**Issue by hsamulowitz** [lmcinnes/umap#535]
Using: umap.UMAP(n_neighbors=15, min_dist=0.1, low_memory=True, random_state=13)
Pynndescent is installed.

Fitting UMAP: umap.fit(train_x)     (a few thousand data points, 10 features)
Applying UMAP transform: umap.transform(test_x) (500 data points, 10 featurs).

UMAP eats up the memory after running umap.tranform() a few thousand times (each time applying it to 500 data points). It gets also very slow. After calling umap.transform() 4k times, the memory usage is at 50GB.

**Comment by dylanking42** [comment]
Could you make a reproducible example? using `sklearn.datasets.make_classification()`  and 
https://stackoverflow.com/questions/938733/total-memory-used-by-python-process
or some other alternative way to report the memory usage

are you storing the results of `umap.transform()` somewhere? without a reproducible example its hard to tell if this is a bug in `UMAP` or in the code you wrote.

**Comment by um-k** [comment]
@hsamulowitz although it looks like `low_memory=True` is by default.

**Comment by lgpreston75** [comment]
I am also having the same problem, even when fit using `low_memory=True`.

```
umap-learn==0.4.6
numba==0.51.2
```

**Comment by rbthomp** [comment]
Has anyone found a solution to this? I'm having the exact same problem. I `fit()` the model then run `transform()` on that model over and over, with each `transform()` growing the memory used. This is problematic when running umap in a web service, with the memory continually growing, the web service periodically has to be restated.

**Comment by giovannigarifo** [comment]
I'm having the same issue on a microservice that uses UMAP, the used memory explodes and the service needs to be restarted, this is also increasing the cost of running the service in a cloud environment. 

Are there plans to solve this problem in a future release? Are there any temporary workarounds available?

**Comment by maryupshall** [comment]
I checked out their docs and it looks like the PyNNDescent library works well with UMAP. Successfully avoided SIGKILL for once. 
here’s the docs: https://pynndescent.readthedocs.io/en/latest/

If you pip or conda install it to your environment UMAP will use it automatically.

**Comment by tomyputw** [comment]
I have a very strange problem. Umap took 10 minutes on a small data set, but when I double the number of columns, the time cost increase to 40 minutes, and the memory usage increase from 10GB to 31GB.

**Comment by maryupshall** [comment]
> I have a very strange problem. Umap took 10 minutes on a small data set, but when I double the number of columns, the time cost increase to 40 minutes, and the memory usage increase from 10GB to 31GB.

Yes, in my experience the memory usage increases exponentially. Try low_memory=True (though this will likely make it slower) and importing pynndescent.
